### PR TITLE
Refactor task acquisition

### PR DESCRIPTION
**Pull request by aneojgurhem**
Aim is to fix https://github.com/aneoconsulting/ArmoniK/issues/266

### CODE REVIEW

File: Adaptors/Memory/src/TaskTable.cs [hunk 1]
```diff
@@ -306,7 +306,7 @@ await grouping.CountAsync(cancellationToken)
   }
 
   /// <inheritdoc />
-  public Task SetTaskErrorAsync(string            taskId,
+  public Task<bool> SetTaskErrorAsync(string            taskId,
```
Comment (lemaitre-aneo) [inline Adaptors/Memory/src/TaskTable.cs:309]: Is this function really synchronous?

File: Adaptors/MongoDB/src/TaskTable.cs [hunk 2]
```diff
@@ -448,43 +454,69 @@ await foreach (var taskId in taskCollection.AsQueryable(sessionHandle)
...
-        Logger.LogInformation("Task {taskId} already acquired by {OtherOwnerPodId}", taskId, ownerPodId);
-        return false;
-      case 1:
-        return true;
-      default:
-        throw new ArmoniKException($"Error during acquisition of task {taskId}");
-    }
+    return res ?? await ReadTaskAsync(taskId,
+                                      cancellationToken)
+             .ConfigureAwait(false);
```
Comment (lemaitre-aneo) [inline Adaptors/MongoDB/src/TaskTable.cs:486]: indentation

File: Adaptors/MongoDB/src/TaskTable.cs [hunk 3]
```diff
@@ -448,43 +454,69 @@ await foreach (var taskId in taskCollection.AsQueryable(sessionHandle)
...
+    using var activity       = activitySource_.StartActivity($"{nameof(ReleaseTask)}");
+    var       taskCollection = taskCollectionProvider_.Get();
+
+    var updateDefinition = new UpdateDefinitionBuilder<TaskData>().Set(tdm => tdm.OwnerPodId,
+                                                                       "")
+                                                                  .Set(tdm => tdm.Status,
+                                                                       TaskStatus.Submitted);
+
+    var filter = new FilterDefinitionBuilder<TaskData>().Where(x => x.TaskId == taskId && x.OwnerPodId == ownerPodId &&
+                                                                    (x.Status == TaskStatus.Processing || x.Status == TaskStatus.Dispatched));
```
Comment (lemaitre-aneo) [inline Adaptors/MongoDB/src/TaskTable.cs:503]: `in` for mongodb request?

File: Adaptors/MongoDB/src/TaskTable.cs [hunk 4]
```diff
@@ -448,43 +454,69 @@ await foreach (var taskId in taskCollection.AsQueryable(sessionHandle)
...
+                                                         new FindOneAndUpdateOptions<TaskData>
+                                                         {
+                                                           ReturnDocument = ReturnDocument.After,
+                                                         },
+                                                         cancellationToken)
+                                  .ConfigureAwait(false);
+
+    return res ?? await ReadTaskAsync(taskId,
+                                      cancellationToken)
+             .ConfigureAwait(false);
```
Comment (lemaitre-aneo) [inline Adaptors/MongoDB/src/TaskTable.cs:519]: indentation

File: Common/src/Pollster/TaskHandler.cs [hunk 5]
```diff
@@ -167,19 +172,30 @@ public async Task<bool> AcquireTask(CancellationToken cancellationToken)
                                                                        cancellationToken)
                                              .ConfigureAwait(false);
 
-        if (!dependencies.Any() || dependencies.Single(i => i.Status == ResultStatus.Completed)
-                                               .Count != taskData_.DataDependencies.Count)
+        if (!dependencies.Any())
         {
-          logger_.LogDebug("Dependencies are not complete yet.");
+          logger_.LogDebug("Dependencies are not available yet.");
```
Comment (lemaitre-aneo) [inline Common/src/Pollster/TaskHandler.cs:177]: "ready"

File: Common/src/Pollster/TaskHandler.cs [hunk 6]
```diff
@@ -214,13 +231,55 @@ public async Task<bool> AcquireTask(CancellationToken cancellationToken)
         return false;
       }
 
-      logger_.LogDebug("checking that the number of retries is not greater than the max retry number");
-      var acquireTask = await taskTable_.AcquireTask(messageHandler_.TaskId,
-                                                     cancellationToken)
-                                        .ConfigureAwait(false);
-      if (!acquireTask)
+      var ownerPodId = LocalIPv4.GetLocalIPv4Ethernet();
```
Comment (lemaitre-aneo) [inline Common/src/Pollster/TaskHandler.cs:234]: cache this

File: Common/src/Pollster/TaskProcessingChecker/TaskProcessingCheckerClient.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    }
+    catch (UriFormatException ex)
+    {
+      logger_.LogWarning(ex,
+                         "Invalid other pod hostname");
+      return false;
+    }
+    catch (Exception ex)
+    {
+      logger_.LogError(ex,
```
Comment (lemaitre-aneo) [inline Common/src/Pollster/TaskProcessingChecker/TaskProcessingCheckerClient.cs:80]: Is it really necessary to log the error here if we rethrow it?

File: Common/tests/ITaskTableTests.cs [hunk 8]
```diff
@@ -722,12 +723,46 @@ public async Task AcquireTaskShouldSucceed()
   {
     if (RunTests)
     {
+      var hostname = Dns.GetHostName();
```
Comment (lemaitre-aneo) [inline Common/tests/ITaskTableTests.cs:726]: `GetLocalIPv4Ethernet`

File: Common/tests/ITaskTableTests.cs [hunk 9]
```diff
@@ -722,12 +723,46 @@ public async Task AcquireTaskShouldSucceed()
...
+                      result!.OwnerPodId);
+    }
+  }
+
+  [Test]
+  public async Task AcquireAcquiredTaskShouldReturnSame()
+  {
+    if (RunTests)
+    {
+      var hostname = Dns.GetHostName();
```
Comment (lemaitre-aneo) [inline Common/tests/ITaskTableTests.cs:745]: `GetLocalIPv4Ethernet`

File: Common/tests/ITaskTableTests.cs [hunk 10]
```diff
@@ -738,10 +773,12 @@ public async Task AcquireTaskShouldFail()
     {
 
       var result = await TaskTable.AcquireTask("TaskFailedId",
+                                               Dns.GetHostName(),
```
Comment (lemaitre-aneo) [inline Common/tests/ITaskTableTests.cs:776]: `GetLocalIPv4Ethernet`

File: Common/tests/ITaskTableTests.cs [hunk 11]
```diff
@@ -738,10 +773,12 @@ public async Task AcquireTaskShouldFail()
...
                                                CancellationToken.None)
                                   .ConfigureAwait(false);
 
-      Assert.IsFalse(result);
+      Assert.AreNotEqual(Dns.GetHostName(),
```
Comment (lemaitre-aneo) [inline Common/tests/ITaskTableTests.cs:780]: `GetLocalIPv4Ethernet`

File: Adaptors/MongoDB/src/TaskTable.cs (hunk 2 above)
Comment (aneojgurhem) [reply Adaptors/MongoDB/src/TaskTable.cs:486]: This is the right indentation according to configuration

File: Adaptors/MongoDB/src/TaskTable.cs (hunk 4 above)
Comment (aneojgurhem) [reply Adaptors/MongoDB/src/TaskTable.cs:519]: This is the right indentation according to configuration

File: Adaptors/MongoDB/src/TaskTable.cs (hunk 3 above)
Comment (lemaitre-aneo) [reply Adaptors/MongoDB/src/TaskTable.cs]: I thought you could just change the part where you test for equality and still use the expression builder.

File: Common/src/Pollster/TaskHandler.cs [hunk 12]
```diff
@@ -214,13 +231,55 @@ public async Task<bool> AcquireTask(CancellationToken cancellationToken)
...
+                               taskData_.OwnerPodId,
+                               taskProcessingElsewhere);
+
+        // if the task is not running on the other pod, we resubmit the task in the queue
+        if (!taskProcessingElsewhere)
+        {
+          taskData_ = await taskTable_.ReadTaskAsync(messageHandler_.TaskId,
+                                                     cancellationToken)
+                                      .ConfigureAwait(false);
+          if (taskData_.OwnerPodId != ownerPodId && taskData_.Status is TaskStatus.Processing or TaskStatus.Dispatched)
```
Comment (lemaitre-aneo) [inline Common/src/Pollster/TaskHandler.cs:262]: Does not need to check for `taskData_.OwnerPodId != ownerPodId`, but would most likely need to check `taskData_.Status is not TaskStatus.Completed and not TaskStatus.Error`

Comment (lemaitre-aneo) [reply Common/src/Pollster/TaskHandler.cs:262]: Or just test for completed (because the other test is done inn resubmit).

**Review by lemaitre-aneo — APPROVED** [review]

**Pull request merged**